src/NuGet.Clients/NuGet.SolutionRestoreManager/InfoBarService.cs

```diff
@@ -0,0 +1,165 @@
+        private bool _closeFromHide = false;
+        private uint? _eventCookie;
+
+        [Import]
+        private Lazy<IPMUIStarter>? PackageManagerUIStarter { get; set; }
+
+        public async Task Show(CancellationToken cancellationToken)
+        {
+            await CreateAndShowInfoBarAsync(cancellationToken);
+        }
+
+        public async Task Hide(CancellationToken cancellationToken)
+        {
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            if (!_visible)
+            {
+                return;
+            }
+
+            _closeFromHide = true;
+
+            try
+            {
+                _infoBarUIElement?.Close();
+                _infoBarUIElement = null;
+            }
+            finally
+            {
+                _closeFromHide = false;
+            }
+        }
+
+        protected async Task<IVsInfoBarHost?> GetInfoBarHostAsync(CancellationToken cancellationToken)
+        {
+            IVsUIShell? uiShell = await _asyncServiceProvider.GetServiceAsync<SVsUIShell, IVsUIShell>(throwOnFailure: false);
+            if (uiShell == null)
+            {
+                return null;
+            }
+
+            // Ensure that we are on the UI thread before interacting with the Solution Explorer UI Element
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            if (ErrorHandler.Failed(uiShell.FindToolWindow((uint)__VSFINDTOOLWIN.FTW_fFindFirst, VSConstants.StandardToolWindows.SolutionExplorer, out var windowFrame)))
+            {
+                return null;
+            }
+
+            object tempObject;
+            if (ErrorHandler.Failed(windowFrame.GetProperty((int)__VSFPROPID7.VSFPROPID_InfoBarHost, out tempObject)))
+            {
+                return null;
+            }
+
+            return tempObject as IVsInfoBarHost;
+        }
+
+        protected InfoBarModel GetInfoBarModel()
+        {
+            return new InfoBarModel(
+                new IVsInfoBarTextSpan[] {
+                    new InfoBarTextSpan(Resources.InfoBar_TextMessage),
+                },
+                new IVsInfoBarActionItem[] {
+                    new InfoBarHyperlink(Resources.InfoBar_HyperlinkMessage),
+                },
+                KnownMonikers.StatusWarning);
+        }
+
+        internal async Task CreateAndShowInfoBarAsync(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (_closed || _visible)
+            {
+                return;
+            }
+
+            try
+            {
+                IVsInfoBarUIFactory? infoBarFactory = await _asyncServiceProvider.GetServiceAsync<SVsInfoBarUIFactory, IVsInfoBarUIFactory>(throwOnFailure: false);
+                if (infoBarFactory == null)
+                {
+                    var exception = new NullReferenceException(nameof(infoBarFactory));
```

Comment:
>var exception = new NullReferenceException(nameof(infoBarFactory)); [](http://example.com/codeflow?start=20&length=67)

For my own learning, how will this Exception be handled?

Reply:
It will be posted to fault teleetry in the next line.

Reply:
Its just for the Telemetry, I'm not throwing anything, should we? In this case the InfoBar will silently fail and not be displayed.

Reply:
I think it's fine to be silent since you have the fault telemetry.